File: largestPlusSign/largestPlusSign.cpp
#include <iostream>
#include <vector>
using namespace std;


class Solution{
public:
int orderOfLargestPlusSign(int N, vector<vector<int>>& mines) {
        int res = 0;
        vector<vector<int> > dp(N, vector<int>(N, N));
        for (int i = 0; i < mines.size(); i++) {
            dp[mines[i][0]][mines[i][1]] = 0;
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0, l = 0; j < N; j++) {
                l = dp[i][j] == 0 ? 0 : l + 1;
                dp[i][j] = min (dp[i][j], l);
            }
            for (int j = N - 1, r = 0; j >= 0; j--) {
                r = dp[i][j] == 0 ? 0 : r + 1;
                dp[i][j] = min (dp[i][j], r);
            }
            for (int j = 0, u = 0; j < N; j++) {
                u = dp[j][i] == 0 ? 0 : u + 1;
                dp[j][i] = min (dp[j][i], u);
            }
            for (int j = N - 1, d = 0; j >= 0; j--) {
                d = dp[j][i] == 0 ? 0 : d + 1;
                dp[j][i] = min (dp[j][i], d);
            }
        }
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                res = max (res, dp[i][j]);
        return res;
    }
};

int main()
{
  Solution s;
  vector<vector<int> > mines = {{4,2}};
  cout << s.orderOfLargestPlusSign(5, mines);
  cout << endl;
  return 0;
}

File: largestPlusSign/largestPlusSign.java
class Solution {
    public int orderOfLargestPlusSign(int n, int[][] mines) {
        int res = 0;
        int[][] grid = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                grid[i][j] = 1;
            }
        }
        for (int[] mine : mines) {
            grid[mine[0]][mine[1]] = 0;
        }
        int[][] left = new int[n][n];
        int[][] right = new int[n][n];
        int[][] up = new int[n][n];
        int[][] down = new int[n][n];
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 0) {
                    right[i][j] = 0;
                    down[i][j] = 0;
                } else {
                    if (j - 1 < 0) right[i][j] = 1;
                    else right[i][j] = right[i][j - 1] + 1;
                    if (i - 1 < 0) down[i][j] = 1;
                    else down[i][j] = down[i - 1][j] + 1;
                }
            }
        }
        
        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (grid[i][j] == 0) {
                    left[i][j] = 0;
                    up[i][j] = 0;
                } else {
                    if (j + 1 >= n) left[i][j] = 1;
                    else left[i][j] = left[i][j + 1] + 1;
                    if (i + 1 >= n) up[i][j] = 1;
                    else up[i][j] = up[i + 1][j] + 1;
                }
            }
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    int c = Math.min(Math.min(up[i][j], down[i][j]), Math.min(left[i][j], right[i][j]));
                    res = Math.max(res, c);
                }
            }
        }
        return res;
    }
}
